Chunk the image horizontally (for landscape images) or vertically (for portrait images) and evaluate faces in each chunk

JPPhoto opened the pull request:
When an image fails initial finding, we now break it into square chunks and attempt the face finding in each chunk.

If all attempts fail, return the whole image as the face-found image so as not to throw an exception.

JPPhoto committed:
Chunk the image horizontally (for landscape images) or vertically (for portrait images) and evaluate faces in each chunk

When an image fails initial finding, we now break it into square chunks and attempt the face finding in each chunk,

JPPhoto committed:
Update faceoff.py

Use a new strategy for finding and picking faces, sorting primarily by x and then y.

JPPhoto committed:
Update faceoff.py

ymgenesis committed:
remove scaling

- added little face_id field to `cleanup_faces_list`
- remove scaling entirely

ymgenesis commented on faceoff.py:
```diff
@@ -29,20 +66,21 @@ class FaceOffOutput(BaseInvocationOutput):
     image:               ImageField  = InputField(description="Image for face detection")
     face_id:             int = InputField(default=0, description="0 for first detected face, single digit for one specific. Multiple faces not supported. Find a face's ID with FaceIdentifier node.")
-    faces:               int = InputField(default=4, description="Maximum number of faces to detect")
+    faces:               int = InputField(default=99, description="Maximum number of faces to detect")
```
Maybe we can just use 10 as a default. It's probably not likely anyone will need more than that and be caught unawares.
```suggestion
    faces:               int = InputField(default=10, description="Maximum number of faces to detect")
```

ymgenesis commented on faceoff.py:
```diff
@@ -63,70 +101,115 @@ def generate_face_box_mask(self, pil_image):
+
+        face_id = self.face_id - 1 if self.face_id > 0 else self.face_id
+
+        if face_id > len(result) or face_id < 0:
+            raise ValueError("Selected face ID is outside of the number of faces detected.")
```
```suggestion

```
This one isn't really raised, can probably remove it and just have a general print statement in the invoke class before passing through the original image.

JPPhoto replied:
I think it's raised if you ask for face 3 and there are only 2, for example. If we don't handle it here, you get an out of bounds index on the array in the next line.

ymgenesis reviewed (commented):
In general it's kind of interesting to pass through the original image when failing... but I guess it's better than stopping the entire graph for now.

I think more testing probably needs to be done with larger portrait images with multiple faces. As mentioned, I'll work on faceidentifier with this implementation and if I can get them to work together we can implement it with faceidentifier and facemask as well.

JPPhoto committed:
Changed number of detected faces default to 10. Removed exceptions and exception handling. Switched to use InvokeAI logging facilities.

ymgenesis committed:
chunk bool, fix face_id catch

- added chunk bool to toggle chunking by default (bypassing full image detection)
- small fix to face_id catch which alerts user that chosen faceID is not in the list of faces.

ymgenesis commented on faceoff.py:
```diff
@@ -77,10 +77,11 @@ class FaceOffInvocation(BaseInvocation):
     def generate_face_box_mask(self, pil_image, chunk_x_offset=0, chunk_y_offset=0):
         result = []
-
+        
```
-_-

ymgenesis committed:
remove whitespace

JPPhoto committed:
Added info logging and bounds checking (with 10% over allowance) to face detection.

JPPhoto committed:
Refactored to separate face detection and returning bounding box

JPPhoto committed:
Fixed two typos.

JPPhoto committed:
Fixed spacing.

JPPhoto committed:
Fixed more spacing issues.

JPPhoto committed:
Fixed bugs, refactored some more

ymgenesis committed:
logging

- Removed a logging line

ymgenesis committed:
remove faces input

- remove faces input so detection is done for all faces, used 999 for mediapipe number of faces to detect
